lib/rom/support/data_proxy.rb

```diff
@@ -0,0 +1,51 @@
+module ROM
+  module DataProxy
+    attr_reader :data, :header, :tuple_proc
+
+    NON_FORWARDABLE = [
+      :each, :to_a, :to_ary, :kind_of?, :instance_of?, :is_a?
+    ].freeze
+
+    def self.included(klass)
+      klass.class_eval do
+        extend ClassMethods
+        include Equalizer.new(:data)
+      end
+    end
+
+    def initialize(data, header, tuple_proc = self.class.tuple_proc)
```

Comment:
I'll just voice my opinion that the name `tuple_proc` doesn't do it for me. I do not have a better name in my head - maybe `row_proc`? Anyway, food for thought as we go forward.